BXMSDOC-5513-master-kogito: Incorporate Edoardo's changes to DRL, XLS, decision engine for Kogito

sterobin opened the pull request:
@evacchi , I've gone through all your changes and have incorporated, with some additional changes to make consistent across the board. I've pinged you in a few places of confusion for confirmation or question.

**Sections especially in need of your careful re-review:**
* [Rule units in DRL](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5513/#con_drl-rule-units_drl-rules) - Also, as usual, I struggled with some aspects of the rule units because I couldn't tell how many rule unit classes you were actually creating. So if you could pay special attention to what I've done, that would be great.
* [Creating DRL rules](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5513/#proc_drl-rules-central-create_drl-rules) - Ditto, tried my best at a few things but still unsure, particularly my final example at the end putting it all together.
* [Metadata tags for fact types](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5513/#ref_drl-declarations-metadata-tags_drl-rules) - Under `@position`, what should the "Example patterns with positional and named arguments" look like in OOPath? @mariofusco ?
* [Property change settings](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5513/#con_property-change-listeners_decision-engine) - Under `@watch`, what should the "Example: Enable or disable property reactivity in fact patterns" look like in OOPath? @mariofusco ? (I see Edoardo said he wasn't sure.)

Links:
* [JIRA](https://issues.redhat.com/browse/BXMSDOC-5513)
* [Enterprise doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5513/#designing_a_decision_service_using_drl_rules)
* [Community doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5513_COMM/#_using_drl_rules_in_kogito_services) - Same, but conditioned

mariofusco commented:
@sterobin I sent a pull request to this with some minor changes, see https://github.com/sterobin/kie-docs/pull/37/files I also added a few comments there, to clarify the reason why I made those changes. @evacchi please also give a look at it.

As for the 2 clarifications you asked, `@position` is not meaningful for oopath (and however almost never used for plain patterns) so I think you can ignore it, while for what regards `@watch` at the moment this haven't been implement yet and we're still discussing on the syntax, so for now you can ignore it as well.

evacchi commented on doc-content/kogito-docs/src/main/asciidoc/drl/con_drl-rule-units.adoc:
```diff
@@ -13,620 +13,263 @@ package org.mortgages;
 ----
 
-Rule units are helpful when you want to coordinate rule execution so that the complete execution of one rule unit triggers the start of another rule unit and so on. For example, assume that you have a set of rules for data enrichment, another set of rules that processes that data, and another set of rules that extract the output from the processed data. If you add these rule sets into three disti ...
+To define a rule unit, you add a DRL file with data type declarations that serves as a rule unit class, and then implement the declared data and data sources using the `RuleUnitData` interface in your DRL rule file, as shown in the following examples:
```
```suggestion
To define a rule unit, you you may use a data type declaration implementing `RuleUnitData` interface.
The data type will declare the data sources as shown in the following examples:
```

evacchi commented on doc-content/kogito-docs/src/main/asciidoc/drl/con_drl-rule-units.adoc:
```diff
@@ -13,620 +13,263 @@ package org.mortgages;
-To override this naming convention and explicitly declare the rule unit that the rules in a DRL file belong to, use the `unit` keyword in the DRL file. The `unit` declaration must immediately follow the package declaration and contain the name of the class in that package that the rules in the DRL file are part of.
-
-.Example rule unit declaration in a DRL file
+.Example DRL rule unit class to define declarations
```
this is not a rule unit. it does not implement RuleUnitData. This is just a type declaration for Person

evacchi commented on doc-content/kogito-docs/src/main/asciidoc/drl/con_drl-rule-units.adoc:
```diff
@@ -13,620 +13,263 @@ package org.mortgages;
-You can also rewrite the same rule condition in a more explicit form using the traditional rule pattern syntax instead of OOPath syntax, as shown in the following example:
-
-.Example rule unit declaration in a DRL file using traditional notation
+.Example DRL rule file to implement rule unit and declarations
```
this DRL declares the rule unit "type" or schema ("declare MortgageRules extends RuleUnitData"), or in other words, it declares the list of data sources using the "declare" syntax;
then there is a rule that uses the included data source declarations.

evacchi commented on doc-content/kogito-docs/src/main/asciidoc/drl/con_drl-rule-units.adoc:
```diff
@@ -2,7 +2,7 @@
 = Rule units in DRL
 
-Rule units are groups of data sources, global variables, and DRL rules that function together for a specific purpose. You can use rule units to partition a rule set into smaller units, bind different data sources to those units, and then execute the individual unit. Rule units are an enhanced alternative to rule-grouping DRL attributes such as rule agenda groups or activation groups for execution ...
+A DRL rule unit is a module for rules and a unit of execution. A rule unit collects a set of rules with the declaration of the type of facts that the rules act on. A rule unit also serves as a unique namespace for each group of rules. A single rule base can contain multiple rule units. You typically store all the rules for a unit in the same file as the unit declaration so that the unit is self-c ...
```
I know that you are trying to go with consistency for the use case, and sorry I introduced a different example here, but the result is a bit confusing. The flow of my original piece:

- first declare data types (alerts and temperatures; in your case Perrson
- then declare the "schema" of the data that will be used, by declaring the data sources
- then declare the rules

evacchi commented on doc-content/kogito-docs/src/main/asciidoc/drl/proc_drl-rules-create.adoc:
```diff
@@ -66,63 +62,24 @@ unit MortgageRules
+import org.kie.kogito.rules.DataStream;
 ----
 --
 * `declare`: (optional) Use this to declare a new fact type to be used by rules in the DRL file. The default fact type in the `java.lang` package of {PRODUCT} is `Object`, but you can declare other types in DRL files as needed. Declaring fact types in DRL files enables you to define a new fact model directly in the {DECISION_ENGINE}, without creating models in a lower-level language like Java.
```
```suggestion
* `declare`: (optional) Use this to declare a new fact type and/or data sources to be used by rules in the DRL file. The default fact type in the `java.lang` package of {PRODUCT} is `Object`, but you can declare other types in DRL files as needed. Declaring fact types in DRL files enables you to define a new fact model directly in the {DECISION_ENGINE}, without creating models in a lower-level language like Java. By implementing `RuleUnitData` you declare the types of data sources to use in a rule unit declaration.
```

sterobin commented:
Okay @evacchi , I've incorporated the feedback from you and @mariofusco , and I've completely redone the "Creating DRL rules" section per our convo.

Can you take on more look at the following to see what you think now? Special attention to terms like "declare" vs. "define" etc, and how I've labeled and explained the snippets, to make sure it's correct now.

Updated sections of focus - **Refresh to get latest**
* [Rule units in DRL](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5513/#con_drl-rule-units_drl-rules) 
* [Creating DRL rules](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5513/#proc_drl-rules-central-create_drl-rules)
